[WIP] Refactor write/read electrodes to a separate function & included coordinate system descriptions

adam2392 opened the pull request:
PR Description
--------------

Closes: #398 

The main portion refactors writing and reading electrodes.tsv files into a `_write_dig_bids` and `_read_dig_bids` functions, and also includes coordinate system descriptions when writing coordsystem.json files. This is a move to help uncomplicate the `read_raw_bids` and `write_raw_bids` functions when dealing w/ electrode coordinates. Eventually, the _write/read_dig_bids funcs could add a public API even?

**Summary**
This is part 1 to address comments at: https://github.com/mne-tools/mne-bids/pull/416#issuecomment-631868195. This will move all related electrodes.tsv and coordsystem.json reading/writing files over to a dig.py. In addition, it makes some minor changes to the naming of the `BIDS_COORDINATE_UNITS`. 

Once the simple "move of functions to dig.py" PR in #425 is merged, these are the things left to do:

- [x] add coordinate system descriptions that are copied from the specification appendix. We can easily accommodate all the coordinate systems accepted by MNE-Python.
- [x] refactor the read section (https://github.com/mne-tools/mne-bids/blob/47ad7ddf8de47e473fa89f4304c044e3406085de/mne_bids/read.py#L554-L555) to a `_read_dig_bids()` function and tidy up the logic 
- [x] refactor the write section (https://github.com/mne-tools/mne-bids/blob/47ad7ddf8de47e473fa89f4304c044e3406085de/mne_bids/write.py#L1257-L1258) to a `_write_dig_bids()` function and tidy up the logic 
- [x] add scalp EEG back into `_write_dig_bids()`, only allowing writing if the landmarks (NAS, LPA, RPA) are present (I'm okay if you want me to table this one, but had this stored from a previous PR, so figured it's easy to add)
- [x] update unit tests 

Merge checklist
---------------

Maintainer, please confirm the following before merging:

- [ ] All comments resolved
- [ ] This is not your own PR
- [ ] All CIs are happy
- [ ] PR title starts with [MRG]
- [ ] [whats_new.rst](https://github.com/mne-tools/mne-bids/blob/master/doc/whats_new.rst) is updated
- [ ] PR description includes phrase "closes <#issue-number>"
- [ ] Commit history does not contain any merge commits

adam2392 commented on mne_bids/tests/test_write.py:
```diff
@@ -752,6 +762,7 @@ def test_edf(_bids_validate):
     # Also cover iEEG
     # We use the same data and pretend that eeg channels are ecog
     ieeg_raw = raw.copy()
+    ieeg_raw.set_montage(None)
```
Needed to add to get rid of the 'captrak' head montage from EEG.

hoechenberger replied:
Could you add this as a comment to the code? :)

adam2392 commented on mne_bids/utils.py:
```diff
@@ -797,3 +797,96 @@ def _scale_coord_to_meters(coord, unit):
         return coord
+
+
+def make_bids_basename(subject=None, session=None, task=None,
```
Moved this here, to prevent circular importing.

adam2392 commented:
@jasmainak lmk what you think? I think the file separation and function separation makes the whole electrodes/coordsystem business a lot cleaner. I added back in the scalp EEG only allowing the `Captrak` coordinate system as per previous discussions.

Added more configuration definitions which were from BIDS specification that we can get for free. E.g. the CoordinateSystemDescription field.

jasmainak commented:
sorry for the delay @adam2392 

@hoechenberger @sappelhoff would you mind taking a first pass? I'll look in a couple of days

jasmainak commented on mne_bids/__init__.py:
```diff
@@ -3,11 +3,11 @@
 from mne_bids.read import (read_raw_bids, get_head_mri_trans,  # noqa: F401
                            get_matched_empty_room)  # noqa: F401
-from mne_bids.utils import make_bids_folders  # noqa: F401
+from mne_bids.utils import make_bids_folders, make_bids_basename  # noqa: F401
```
yes, this is exactly what we should be doing! cc @hoechenberger

jasmainak commented on mne_bids/dig.py:
```diff
@@ -0,0 +1,427 @@
+
+def _electrodes_tsv(raw, fname, kind, overwrite=False, verbose=True):
+    """
+    Create an electrodes.tsv file and save it.
```
PEP8 convention. This line should go above

jasmainak commented on mne_bids/dig.py:
```diff
@@ -0,0 +1,427 @@
+    fname : str
+        Filename to save the electrodes.tsv to.
+    kind : str
+        Acquisition type to save electrodes in. For iEEG, requires size.
```
humm ... where will get this from?

sappelhoff replied:
also, "kind" is often rather described as the "data modality" (eeg, ieeg, meg, ...) ... the word "acquisition" is too reminiscent of the `acq` _entity_ which does a different thing than the `kind` _suffix_

jasmainak commented on mne_bids/dig.py:
```diff
@@ -0,0 +1,427 @@
+    raw : instance of Raw
+        The data as MNE-Python Raw object.
+    unit : str
+        Units to be used in the coordsystem specification.
```
What are the valid choices?

jasmainak commented:
@adam2392 the diff is huge. Part of the reason is that you have moved things around. Github doesn't do a good job of tracking changes when you move things. Plus it's difficult to review. 

What I suggest is to break down the PR into two parts

1. Just move things around to make it easier to make concrete API changes.
2. Do the actual refactoring and API changes.

Ideally, 1. can be merged without much review and for 2., we will do the actual review. Do you think this is possible?

Also it would be really helpful if you can summarize in a concise comment what you moved where and any other changes that needed to be done.

adam2392 commented:
> Also it would be really helpful if you can summarize in a concise comment what you moved where and any other changes that needed to be done.

@jasmainak 

**Summary**
This is part 1 to address comments at: https://github.com/mne-tools/mne-bids/pull/416#issuecomment-631868195. This will move all related electrodes.tsv and coordsystem.json reading/writing files over to a dig.py. In addition, it makes some minor changes to the naming of the `BIDS_COORDINATE_UNITS`. 

Once the simple "move of functions to dig.py" PR in #425 is merged, these are the things left to do:

- [ ] add coordinate system descriptions that are copied from the specification appendix. We can easily accommodate all the coordinate systems accepted by MNE-Python.
- [ ] refactor the read section (https://github.com/mne-tools/mne-bids/blob/47ad7ddf8de47e473fa89f4304c044e3406085de/mne_bids/read.py#L554-L615) to a `_read_dig_bids()` function and tidy up the logic 
- [ ] refactor the write section (https://github.com/mne-tools/mne-bids/blob/47ad7ddf8de47e473fa89f4304c044e3406085de/mne_bids/write.py#L1257-L1296) to a `_write_dig_bids()` function and tidy up the logic 
- [ ] add scalp EEG back into `_write_dig_bids()`, only allowing writing if the landmarks (NAS, LPA, RPA) are present (I'm okay if you want me to table this one, but had this stored from a previous PR, so figured it's easy to add)
- [ ] update unit tests

agramfort commented:
@adam2392 I am lost. What should get in first? what is good to go from your end?

adam2392 commented:
> @adam2392 I am lost. What should get in first? what is good to go from your end?

Sorry first time making such a large change...

#425 should go in first according to @jasmainak comments. This PR is just moving the existing code out of write.py and read.py. The next PR will address the remaining check list.

jasmainak commented:
@adam2392 sounds like a plan, please update the PR description so that this is not lost in a comment.

sappelhoff commented on mne_bids/config.py:
```diff
@@ -57,24 +57,78 @@
+BIDS_MEG_COORDINATE_FRAMES = ['ctf', 'elektaneuromag',
+                              '4dbti', 'kityokogawa',
+                              'chietiitab', 'other']
+BIDS_EEG_COORDINATE_FRAMES = ['captrak']  # XXX: We don't include BESA
```
```suggestion
BIDS_EEG_COORDINATE_FRAMES = ['captrak']
```

BESA has been completely removed now as a bugfix, we don't even need to mention it

sappelhoff commented on mne_bids/dig.py:
```diff
@@ -188,23 +220,219 @@ def _coordsystem_json(raw, unit, orient, coordsystem_name, fname,
+        Filename to save the electrodes.tsv to.
+    coordsystem_fname : str
+        Filename to save the coordsystem.json to.
+    data_path : str | | pathlib.Path
```
```suggestion
    data_path : str | pathlib.Path
```

sappelhoff commented on mne_bids/dig.py:
```diff
@@ -188,23 +220,219 @@ def _coordsystem_json(raw, unit, orient, coordsystem_name, fname,
+
+        if verbose:
+            print("Writing electrodes file to... ", electrodes_fname)
+            print("Writing coordsytem file to... ", coordsystem_fname)
```
```suggestion
            print("Writing coordsystem file to... ", coordsystem_fname)
```

sappelhoff commented on mne_bids/tests/test_write.py:
```diff
@@ -781,7 +801,7 @@ def test_edf(_bids_validate):
     with open(coordsystem_fname, 'r') as fin:
         coordsystem_json = json.load(fin)
-    assert coordsystem_json['iEEGCoordinateSystem'] == 'Other'
+    assert coordsystem_json['iEEGCoordinateSystem'] == 'other'
```
The `other` keyword should be with a capital `O` according to the BIDS spec: https://bids-specification.readthedocs.io/en/stable/99-appendices/08-coordinate-systems.html#coordinate-systems-applicable-to-meg-eeg-and-ieeg

this probably has to be fixed in some part of the code?

adam2392 replied:
Done

sappelhoff commented on mne_bids/config.py:
```diff
@@ -57,24 +57,78 @@
+MNE_FRAME_TO_STR = {val: key for key, val in MNE_STR_TO_FRAME.items()}
+
+# see BIDS specification for description we copied over from each
+COORD_FRAME_DESCRIPTIONS = {
```
did I miss it or are these not used anywhere?

adam2392 replied:
These are used in `_coordsystem_json()`... which I"m looking now and it did not rebase over :(. I'll add those back in manually.

FYI, there's no great unit test for this except checking the coordsystem_json file manually, but I think it's a pretty free feature to add, w/o any downstream dependencies.

sappelhoff commented on mne_bids/dig.py:
```diff
@@ -188,23 +220,219 @@ def _coordsystem_json(raw, unit, orient, coordsystem_name, fname,
+           for _digpoint in raw.info['dig']):
+        warn("Not all digpoints have the same coordinate frame. "
+             "Skipping electrodes.tsv writing...")
+    else:
```
you can also just put a `return` after the warning and get rid of the `else` nesting right below. "flat is better than nested" (python zen)

sappelhoff commented on mne_bids/dig.py:
```diff
@@ -188,23 +220,219 @@ def _coordsystem_json(raw, unit, orient, coordsystem_name, fname,
+
+            # XXX: to be improved,
+            # mne-python automatically converts unknown coord frame to head
+            if coord_frame_int == FIFF.FIFFV_COORD_HEAD and landmarks:
```
Here I would write either
- if coord_frame is HEAD
- or if we have all landmarks (then we rescale to head)

do you agree with this? Does the `XXX` just mean that this approach is currently not possible? what needs to change to make it possible?

adam2392 replied:
Doesn't mne-python rescale to head automatically? The XXX is to allow rescaling if there exist landmarks + coordsystem is 'unknown'?

Idk scalp EEG montage that well.

adam2392 replied:
I have that issue too in iEEG data, so here was my proposed temporary fix, but I think @agramfort and others chimed in about the difficulties of this since a lot of downstream functionality depends on it. 

https://github.com/mne-tools/mne-python/pull/7709

sappelhoff commented:
also @adam2392 does this fix https://github.com/mne-tools/mne-bids/issues/203 ?

adam2392 commented:
> also @adam2392 does this fix #203 ?

Addressed all the comments to the best I could. Regarding #203 , nope but it can in the future if I'm understanding correctly.

sappelhoff commented on mne_bids/config.py:
```diff
@@ -57,24 +57,78 @@
+    '4dbti': 'ALS orientation and the origin between the ears',
+    'kityokogawa': 'ALS orientation and the origin between the ears',
+    'chietiitab': 'RAS orientation and the origin between the ears',
+    'besa': 'Although natively this is a spherical coordinate system, '
+            'the electrode positions should be expressed in '
+            'Cartesian coordinates, with a RAS orientation. '
+            'The X axis is the T8-T7 line, positive at T8. '
+            'The Y axis is the Oz-Fpz line, positive at Fpz. '
+            'The origin of the sphere fitted to the electrodes '
+            'is approximately 4 cm above the point between the ears.',
```
```suggestion
```

sappelhoff commented on mne_bids/dig.py:
```diff
@@ -182,29 +225,233 @@ def _coordsystem_json(raw, unit, orient, coordsystem_name, fname,
 
+    # get the coordinate frame description
+    coordsystem_desc = COORD_FRAME_DESCRIPTIONS.get(coordsystem_name, "n/a")
+
+    # create the coordinate json data structure based on 'kind'
```
I think it'd be good to include a log message here (not a warning, just a message that can be handled with verbose)

it should say that if your coordsystem_name is `Other` ... you HAVE TO manually define the CoordsystemDescription in your dataset later on.

Because the whole purpose of BIDS is defeated it people put "other" somewhere and never describe what it is.
```suggestion
    if coordsystem_name == 'Other' and verbose:
        print('Using the `Other` keyword for the CoordinateSystem field. '
              'Please specify the CoordinateSystemDescription field manually.')
    
    # create the coordinate json data structure based on 'kind'
```

sappelhoff commented on mne_bids/dig.py:
```diff
@@ -182,29 +225,233 @@ def _coordsystem_json(raw, unit, orient, coordsystem_name, fname,
+        coords = _extract_landmarks(raw.info['dig'])
+        landmarks = set(['RPA', 'NAS', 'LPA']) == set(list(coords.keys()))
+
+        # XXX: to be improved to allow rescaling if landmarks
```
```suggestion
        # XXX: to be improved to allow rescaling if landmarks are present
```

agramfort commented:
thx @adam2392 

we're making fast progress here !
